bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java

```diff
@@ -90,137 +85,82 @@ protected void deactivate(ComponentContext componentContext) {
         }
         thingTypeRegistryServiceTracker.close();
         configDescriptionRegistryServiceTracker.close();
-        this.thingHandlers.clear();
-        this.configStatusProviders.clear();
-        this.firmwareUpdateHandlers.clear();
-        this.bundleContext = null;
+        configStatusProviders.clear();
+        firmwareUpdateHandlers.clear();
+        bundleContext = null;
     }
 
     @Override
-    public void unregisterHandler(Thing thing) {
-        ServiceRegistration<ThingHandler> thingHandlerServiceRegistration = thingHandlers
-                .remove(thing.getUID().toString());
-        if (thingHandlerServiceRegistration != null) {
-            logger.debug("Removing service registration for handler of thing '{}'.", thing.getUID().getAsString());
-            unregisterHandler(thingHandlerServiceRegistration);
-        } else {
-            logger.error("There is no service registration for a thing handler of thing '{}'.",
-                    thing.getUID().getAsString());
-        }
-        ServiceRegistration<ConfigStatusProvider> configStatusProviderServiceRegistration = configStatusProviders
-                .remove(thing.getUID().getAsString());
-        if (configStatusProviderServiceRegistration != null) {
-            configStatusProviderServiceRegistration.unregister();
-        }
-        ServiceRegistration<FirmwareUpdateHandler> firmwareUpdateHandlerServiceRegistration = firmwareUpdateHandlers
-                .remove(thing.getUID().getAsString());
-        if (firmwareUpdateHandlerServiceRegistration != null) {
-            firmwareUpdateHandlerServiceRegistration.unregister();
-        }
-    }
+    public ThingHandler registerHandler(Thing thing) {
+        Preconditions.checkArgument(thing != null, "The argument 'thing' must not be null.");
 
-    private void unregisterHandler(ServiceRegistration<ThingHandler> serviceRegistration) {
-        ThingHandler thingHandler = bundleContext.getService(serviceRegistration.getReference());
-        removeHandler(thingHandler);
-        serviceRegistration.unregister();
-        if (thingHandler instanceof BaseThingHandler) {
-            ((BaseThingHandler) thingHandler).unsetBundleContext(bundleContext);
-        }
-    }
-
-    @Override
-    public void registerHandler(Thing thing, ThingHandlerCallback thingHandlerListener) {
         ThingHandler thingHandler = createHandler(thing);
         if (thingHandler == null) {
             throw new IllegalStateException(this.getClass().getSimpleName()
                     + " could not create a handler for the thing '" + thing.getUID() + "'.");
         }
+        setHandlerContext(thingHandler);
```

Comment:
> BundleContext, ThingRegistry, ItemChannelLinkRegistry would be removed from the BaseThingHandler.

This imho makes sense as these services should not be directly available to binding developers.

> This will break a lot of bindings

I hope that it doesn't affect to many bindings after all - from a quick check in the ESH+OH2 repos I only found 4 that would need to be changed.

> I would like to provide this refactoring in a separate PR.

If this is something that is pretty isolated from the other changes, doing it in a separate PR is fine for me - I'd just like to have your confirmation that you'll nonetheless continue to work on it :-)